Sources/Async/Streams/ByteParserStream.swift

```diff
@@ -0,0 +1,70 @@
+public protocol ByteParserStream: TranslatingStream where Input == UnsafeBufferPointer<UInt8> {
+    /// Any type that can be used to carry partially completed data into `continueParsing`
+    ///
+    /// If there are multiple partial states you can use an enum
+    associatedtype Partial
+    
+    /// A state kept by the Async library used to keep track of which data is parsed and if more data is needed
+    var state: ByteParserStreamState<Self> { get }
+    
+    /// Continues parsing a partially parsed Output
+    func continueParsing(_ partial: Partial, from buffer: Input) throws -> ByteParserResult<Partial, Output>
+    
+    /// Starts parsing Output, resulting in either a partially or fully completed Output
+    func startParsing(from buffer: Input) throws -> ByteParserResult<Partial, Output>
+}
+
+public final class ByteParserStreamState<S: ByteParserStream> {
+    /// The current eventloop, used to dispatch tasks (preventing stack overflows)
+    fileprivate var eventloop: EventLoop
+    
+    /// The current offset where is being parsed
+    fileprivate var parsedInput: Int
+    
+    /// Stores partially parsed data
+    fileprivate var partiallyParsed: S.Partial?
+    
+    public init(worker: Worker) {
+        eventloop = worker.eventLoop
+        parsedInput = 0
+    }
+}
+
+public enum ByteParserResult<Partial, Output> {
```

Comment:
this should probably just be generic on the byte parser stream as well. more future proof that way